urllib3/pipeline/base.py

```diff
@@ -0,0 +1,145 @@
+from uuid import uuid4
+
+import six
+
+from urllib3.pipeline.context import Context, ContextWrapper
+
+def has_callable_attr(obj, attr_name):
+    return hasattr(obj, attr_name) and hasattr(getattr(obj, attr_name), '__call__')
+
+
+class HttpPipelineElement(object):
+    
+    def __init__(self):
+        """
+        Each instance of HttpPipelineElement should have a unique identifier
+        that can be used to store state in a context object.
+        """
+        self.id = uuid4().hex
+
+    def save_context(self, context, key, value):
+
+        context.setdefault(self.id, {})
+        context[self.id][key] = value
+
+    def get_context(self, context, key, default=None):
+        return context.get(self.id, {}).get(key, default)
```

Comment:
This logic makes it impossible to store a `None` value in the context. Do we want that limitation?